Sources/performance-worker/WorkerService.swift

```diff
@@ -114,98 +68,176 @@ final class WorkerService: Grpc_Testing_WorkerService.ServiceProtocol, Sendable
+        for client in state.clients {
+          try? state.rpcStats.merge(client.currentStats)
+        }
+
+        self.role = .client(state)
+        return (stats, state.rpcStats)
+
       case .server, .none:
         return nil
       }
     }
 
-    mutating func setupServer(server: GRPCServer, stats: ServerStats) throws {
-      let serverState = State.ServerState(server: server, stats: stats)
-      switch self.role {
-      case .server(_):
-        throw RPCError(code: .alreadyExists, message: "A server has already been set up.")
+    enum OnStartedServer {
+      case runServer
+      case invalidState(RPCError)
+    }
 
-      case .client(_):
-        throw RPCError(code: .failedPrecondition, message: "This worker has a client setup.")
+    mutating func startedServer(
+      _ server: GRPCServer,
+      stats: ServerStats,
+      eventLoopGroup: MultiThreadedEventLoopGroup
+    ) -> OnStartedServer {
+      let action: OnStartedServer
 
+      switch self.role {
       case .none:
-        self.role = .server(serverState)
+        let state = State.Server(server: server, stats: stats, eventLoopGroup: eventLoopGroup)
+        self.role = .server(state)
+        action = .runServer
+      case .server:
+        let error = RPCError(code: .alreadyExists, message: "A server has already been set up.")
+        action = .invalidState(error)
+      case .client:
+        let error = RPCError(code: .failedPrecondition, message: "This worker has a client setup.")
+        action = .invalidState(error)
       }
+
+      return action
     }
 
-    mutating func setupClients(
-      benchmarkClients: [BenchmarkClient],
+    enum OnStartedClients {
+      case runClients
+      case invalidState(RPCError)
+    }
+
+    mutating func startedClients(
+      _ clients: [BenchmarkClient],
       stats: ClientStats,
       rpcStats: RPCStats
-    ) throws {
-      let clientState = State.ClientState(
-        clients: benchmarkClients,
-        stats: stats,
-        rpcStats: rpcStats
-      )
+    ) -> OnStartedClients {
+      let action: OnStartedClients
+
       switch self.role {
-      case .server(_):
-        throw RPCError(code: .alreadyExists, message: "This worker has a server setup.")
+      case .none:
+        let state = State.Client(clients: clients, stats: stats, rpcStats: rpcStats)
+        self.role = .client(state)
+        action = .runClients
+      case .server:
+        let error = RPCError(code: .alreadyExists, message: "This worker has a server setup.")
+        action = .invalidState(error)
+      case .client:
+        let error = RPCError(
+          code: .failedPrecondition,
+          message: "Clients have already been set up."
+        )
+        action = .invalidState(error)
+      }
 
-      case .client(_):
-        throw RPCError(code: .failedPrecondition, message: "Clients have already been set up.")
+      return action
+    }
 
+    mutating func serverShutdown() -> EventLoopGroup? {
```

Comment:
Bit of a nit, but should we have an action with cases `shutdownEventLoopGroup(EventLoopGroup)` and `nothing` instead of this optional ELG? It would be more consistent with other funcs and also clearer about what we're supposed to do with the ELG).